diff --git a/examples/tutorials/bottle.py b/examples/tutorials/bottle.py
new file mode 100644
index 0000000..105513d
--- /dev/null
+++ b/examples/tutorials/bottle.py
@@ -0,0 +1,86 @@
+import genesis as gs
+import torch
+
+########################## init ##########################
+gs.init()
+
+########################## create a scene ##########################
+
+scene = gs.Scene(
+    sim_options=gs.options.SimOptions(
+        dt=1e-2,
+        substeps=100,
+    ),
+    mpm_options=gs.options.MPMOptions(
+        lower_bound=(-1.0, -1.0, -1.0),
+        upper_bound=(1.0, 1.0, 1),
+    ),
+    vis_options=gs.options.VisOptions(
+        visualize_mpm_boundary=True,
+    ),
+    viewer_options=gs.options.ViewerOptions(
+        camera_fov=30,
+        res=(960, 640),
+    ),
+    show_viewer=True,
+)
+
+########################## entities ##########################
+plane = scene.add_entity(
+    morph=gs.morphs.Plane(pos=(-1, -1, -1)),
+)
+
+
+import os
+materials = {
+    "base": gs.materials.MPM.Base,
+    "elastic": gs.materials.MPM.Elastic,
+    "elastoplastic": gs.materials.MPM.ElastoPlastic,
+    "liquid": gs.materials.MPM.Liquid,
+    "muscle": gs.materials.MPM.Muscle,
+    "sand": gs.materials.MPM.Sand,
+    "snow": gs.materials.MPM.Snow,
+}
+
+MATERIAL="elastic"
+MODEL_PATH="output/bottle"
+OBJ_PATH=os.path.join(MODEL_PATH, "pseudomesh_info/ours_30000/scale_1.obj")
+SAVE_PATH=os.path.join(MODEL_PATH, "genesis_triangles", MATERIAL)
+os.makedirs(SAVE_PATH, exist_ok=True)
+SCALE=0.5
+mesh = scene.add_entity(
+    material=materials[MATERIAL](sampler="regular"),#(E=1e5, nu=0.1, rho=1000),
+    morph=gs.morphs.Mesh(file=OBJ_PATH, convexify=False, decompose_nonconvex=False, scale=SCALE)
+)
+
+scene.build()
+
+########################## build ##########################
+
+
+# def calc_scales(pts):
+#     if isinstance(pts, list):
+#         pts = torch.concatenate(pts)
+#     x = pts
+#     scales = x.reshape((-1, 3, 3))
+#     scales = scales - scales[:, 0, :].unsqueeze(-2)
+#     scales = torch.linalg.norm(scales, axis=-1)
+#     return scales.reshape(-1, 1)
+
+# def clamp_pos(x, init_scales, threshold=1.5):
+#     x = x.cpu()
+#     scales = calc_scales(x).expand(-1, 3)
+#     return torch.where(scales > threshold * init_scales, x + threshold * init_scales, x).cuda()
+
+
+# init_scales = calc_scales(mesh.get_state().pos[mesh.particle_start:mesh.particle_end]).expand(-1, 3).cpu()
+
+horizon = 150
+for i in range(horizon + 1):
+    scene.step()
+    # if i % 2 == 0:
+        # mesh.get_state().pos[mesh.particle_start:mesh.particle_end] = clamp_pos(mesh.get_state().pos[mesh.particle_start:mesh.particle_end], init_scales).cuda()
+    torch.save(
+        torch.tensor(mesh.get_state().pos[mesh.particle_start:mesh.particle_end].clone().detach().to("cpu")).reshape(-1, 3, 3),
+        os.path.join(SAVE_PATH, f"{i:05d}.pt")
+    )
diff --git a/examples/tutorials/mpm.py b/examples/tutorials/mpm.py
index 1c6a386..49be9b3 100644
--- a/examples/tutorials/mpm.py
+++ b/examples/tutorials/mpm.py
@@ -1,4 +1,5 @@
 import genesis as gs
+import torch
 
 ########################## init ##########################
 gs.init()
@@ -11,8 +12,10 @@ scene = gs.Scene(
         substeps=10,
     ),
     mpm_options=gs.options.MPMOptions(
-        lower_bound=(-0.5, -1.0, 0.0),
-        upper_bound=(0.5, 1.0, 1),
+        # lower_bound=(-0.5, -1.0, 0.0),
+        # upper_bound=(0.5, 1.0, 1),
+        lower_bound=(-1.0, -1.0, -1.0),
+        upper_bound=(1.0, 1.0, 1),
     ),
     vis_options=gs.options.VisOptions(
         visualize_mpm_boundary=True,
@@ -26,49 +29,40 @@ scene = gs.Scene(
 
 ########################## entities ##########################
 plane = scene.add_entity(
-    morph=gs.morphs.Plane(),
+    morph=gs.morphs.Plane(pos=(-1, -1, -1)),
 )
 
-obj_elastic = scene.add_entity(
-    material=gs.materials.MPM.Elastic(),
-    morph=gs.morphs.Box(
-        pos=(0.0, -0.5, 0.25),
-        size=(0.2, 0.2, 0.2),
-    ),
-    surface=gs.surfaces.Default(
-        color=(1.0, 0.4, 0.4),
-        vis_mode="visual",
-    ),
-)
 
-obj_sand = scene.add_entity(
-    material=gs.materials.MPM.Liquid(),
-    morph=gs.morphs.Box(
-        pos=(0.0, 0.0, 0.25),
-        size=(0.3, 0.3, 0.3),
-    ),
-    surface=gs.surfaces.Default(
-        color=(0.3, 0.3, 1.0),
-        vis_mode="particle",
-    ),
-)
+import os
+materials = {
+    "base": gs.materials.MPM.Base,
+    "elastic": gs.materials.MPM.Elastic,
+    "elastoplastic": gs.materials.MPM.ElastoPlastic,
+    "liquid": gs.materials.MPM.Liquid,
+    "muscle": gs.materials.MPM.Muscle,
+    "sand": gs.materials.MPM.Sand,
+    "snow": gs.materials.MPM.Snow,
+}
 
-obj_plastic = scene.add_entity(
-    material=gs.materials.MPM.ElastoPlastic(),
-    morph=gs.morphs.Sphere(
-        pos=(0.0, 0.5, 0.35),
-        radius=0.1,
-    ),
-    surface=gs.surfaces.Default(
-        color=(0.4, 1.0, 0.4),
-        vis_mode="particle",
-    ),
+MATERIAL="elastic"
+MODEL_PATH="output\\bottle"
+OBJ_PATH=os.path.join(MODEL_PATH, "pseudomesh_info\\ours_30000\\scale_1.obj")
+SAVE_PATH=os.path.join(MODEL_PATH, "genesis_triangles", MATERIAL)
+os.makedirs(SAVE_PATH, exist_ok=True)
+SCALE=0.4
+mesh = scene.add_entity(
+    material=gs.materials.MPM.Elastic(sampler="random"),#(E=1e5, nu=0.1, rho=1000),
+    morph=gs.morphs.Mesh(file=OBJ_PATH, convexify=False, decompose_nonconvex=False, scale=SCALE)
 )
 
+scene.build()
 
 ########################## build ##########################
-scene.build()
 
-horizon = 1000
+horizon = 200
 for i in range(horizon):
     scene.step()
+    saved_tensor = torch.tensor(mesh.get_state().pos[mesh.particle_start:mesh.particle_end].clone().detach().to("cpu")).reshape(-1, 3, 3)
+    torch.save(saved_tensor, os.path.join(SAVE_PATH, f"{i:05d}.pt"))
+    del saved_tensor
+    torch.cuda.empty_cache()
diff --git a/examples/tutorials/teddybear.py b/examples/tutorials/teddybear.py
new file mode 100644
index 0000000..109873a
--- /dev/null
+++ b/examples/tutorials/teddybear.py
@@ -0,0 +1,86 @@
+import genesis as gs
+import torch
+
+########################## init ##########################
+gs.init()
+
+########################## create a scene ##########################
+
+scene = gs.Scene(
+    sim_options=gs.options.SimOptions(
+        dt=1e-2,
+        substeps=100,
+    ),
+    mpm_options=gs.options.MPMOptions(
+        lower_bound=(-2.0, -2.0, -2.0),
+        upper_bound=(2.0, 2.0, 2),
+    ),
+    vis_options=gs.options.VisOptions(
+        visualize_mpm_boundary=True,
+    ),
+    viewer_options=gs.options.ViewerOptions(
+        camera_fov=30,
+        res=(960, 640),
+    ),
+    show_viewer=True,
+)
+
+########################## entities ##########################
+plane = scene.add_entity(
+    morph=gs.morphs.Plane(pos=(-1, -1, -1)),
+)
+
+
+import os
+materials = {
+    "base": gs.materials.MPM.Base,
+    "elastic": gs.materials.MPM.Elastic,
+    "elastoplastic": gs.materials.MPM.ElastoPlastic,
+    "liquid": gs.materials.MPM.Liquid,
+    "muscle": gs.materials.MPM.Muscle,
+    "sand": gs.materials.MPM.Sand,
+    "snow": gs.materials.MPM.Snow,
+}
+
+MATERIAL="sand"
+MODEL_PATH="output\\teddybear"
+OBJ_PATH=os.path.join(MODEL_PATH, "pseudomesh_info\\ours_30000\\scale_1.obj")
+SAVE_PATH=os.path.join(MODEL_PATH, "genesis_triangles", MATERIAL)
+os.makedirs(SAVE_PATH, exist_ok=True)
+SCALE=0.5
+mesh = scene.add_entity(
+    material=materials[MATERIAL](sampler="random"),#(E=1e5, nu=0.1, rho=1000),
+    morph=gs.morphs.Mesh(file=OBJ_PATH, convexify=False, decompose_nonconvex=False, scale=SCALE)
+)
+
+scene.build()
+
+########################## build ##########################
+
+
+# def calc_scales(pts):
+#     if isinstance(pts, list):
+#         pts = torch.concatenate(pts)
+#     x = pts
+#     scales = x.reshape((-1, 3, 3))
+#     scales = scales - scales[:, 0, :].unsqueeze(-2)
+#     scales = torch.linalg.norm(scales, axis=-1)
+#     return scales.reshape(-1, 1)
+
+# def clamp_pos(x, init_scales, threshold=1.5):
+#     x = x.cpu()
+#     scales = calc_scales(x).expand(-1, 3)
+#     return torch.where(scales > threshold * init_scales, x + threshold * init_scales, x).cuda()
+
+
+# init_scales = calc_scales(mesh.get_state().pos[mesh.particle_start:mesh.particle_end]).expand(-1, 3).cpu()
+
+horizon = 150
+for i in range(horizon + 1):
+    scene.step()
+    # if i % 2 == 0:
+        # mesh.get_state().pos[mesh.particle_start:mesh.particle_end] = clamp_pos(mesh.get_state().pos[mesh.particle_start:mesh.particle_end], init_scales).cuda()
+    torch.save(
+        torch.tensor(mesh.get_state().pos[mesh.particle_start:mesh.particle_end].clone().detach().to("cpu")).reshape(-1, 3, 3) / SCALE,
+        os.path.join(SAVE_PATH, f"{i:05d}.pt")
+    )
diff --git a/genesis/utils/particle.py b/genesis/utils/particle.py
index aa45e9b..4cc9b82 100644
--- a/genesis/utils/particle.py
+++ b/genesis/utils/particle.py
@@ -66,11 +66,11 @@ def trimesh_to_particles_simple(mesh, p_size, sampler):
     if not os.path.exists(ptc_file_path):
         with gs.logger.timer(f"Sampling particles with ~<{sampler}>~ sampler and generating `.ptc` file:"):
             # sample a cube first
-            box_size = mesh.bounds[1] - mesh.bounds[0]
-            box_center = (mesh.bounds[1] + mesh.bounds[0]) / 2
-            positions = _box_to_particles(p_size=p_size, pos=box_center, size=box_size, sampler=sampler)
+            # box_size = mesh.bounds[1] - mesh.bounds[0]
+            # box_center = (mesh.bounds[1] + mesh.bounds[0]) / 2
+            positions = mesh.vertices #_box_to_particles(p_size=p_size, pos=box_center, size=box_size, sampler=sampler)
             # reject out-of-boundary particles
-            positions = positions[igl.signed_distance(positions, mesh.vertices, mesh.faces)[0] < 0]
+            #positions = positions[igl.signed_distance(positions, mesh.vertices, mesh.faces)[0] < 0]
 
             os.makedirs(os.path.dirname(ptc_file_path), exist_ok=True)
             pkl.dump(positions, open(ptc_file_path, "wb"))
